reactive/calico.py

```diff
@@ -21,19 +26,96 @@
 ETCD_CERT_PATH = os.path.join(CALICOCTL_PATH, 'etcd-cert')
 ETCD_CA_PATH = os.path.join(CALICOCTL_PATH, 'etcd-ca')
 CALICO_CIDR = '192.168.0.0/16'
+CALICO_UPGRADE_DIR = '/opt/calico-upgrade'
 
 
 @hook('upgrade-charm')
 def upgrade_charm():
     remove_state('calico.binaries.installed')
     remove_state('calico.cni.configured')
+    remove_state('calico.service.installed')
+    remove_state('calico.pool.configured')
+    remove_state('calico.cni.configured')
+    remove_state('calico.npc.deployed')
     try:
         log('Deleting /etc/cni/net.d/10-calico.conf')
         os.remove('/etc/cni/net.d/10-calico.conf')
     except FileNotFoundError as e:
         log(e)
+    if is_leader() and not leader_get('calico-v3-data-ready'):
+        leader_set({
+            'calico-v3-data-migration-needed': True,
+            'calico-v3-npc-cleanup-needed': True,
+            'calico-v3-completion-needed': True
+        })
+
+
+@when('leadership.is_leader', 'leadership.set.calico-v3-data-migration-needed',
+      'etcd.available', 'calico.etcd-credentials.installed')
+def upgrade_v3_migrate_data():
+    status_set('maintenance', 'Migrating data to Calico 3')
+    try:
+        calico_upgrade.configure()
+        calico_upgrade.dry_run()
+        calico_upgrade.start()
+    except Exception:
+        log(traceback.format_exc())
+        message = 'Calico upgrade failed, see debug log'
+        status_set('blocked', message)
+        return
+    leader_set({'calico-v3-data-migration-needed': None})
+
+
+@when('leadership.is_leader')
+@when_not('leadership.set.calico-v3-data-migration-needed')
+def v3_data_ready():
+    leader_set({'calico-v3-data-ready': True})
+
+
+@when('leadership.is_leader', 'leadership.set.calico-v3-data-ready',
+      'leadership.set.calico-v3-npc-cleanup-needed')
+def upgrade_v3_npc_cleanup():
+    status_set('maintenance', 'Cleaning up Calico 2 policy controller')
+
+    resources = [
+        ('Deployment', 'kube-system', 'calico-policy-controller'),
+        ('ClusterRoleBinding', None, 'calico-policy-controller'),
+        ('ClusterRole', None, 'calico-policy-controller'),
+        ('ServiceAccount', 'kube-system', 'calico-policy-controller')
+    ]
+
+    for kind, namespace, name in resources:
+        args = ['delete', '--ignore-not-found', kind, name]
+        if namespace:
+            args += ['-n', namespace]
+        try:
+            kubectl(*args)
+        except CalledProcessError:
+            log('Failed to cleanup %s %s %s' % (kind, namespace, name))
+            return
+
+    leader_set({'calico-v3-npc-cleanup-needed': None})
+
+
+@when('leadership.is_leader', 'leadership.set.calico-v3-completion-needed',
+      'leadership.set.calico-v3-data-ready', 'calico.binaries.installed',
+      'calico.service.installed', 'calico.npc.deployed')
+@when_not('leadership.set.calico-v3-npc-cleanup-needed')
```

Comment:
We don't need to worry about data migration being None here?

Reply:
`leadership.set.calico-v3-data-ready` implies `not leadership.set.calico-v3-data-migration-needed`. We don't need both, but I can add the latter if it makes sense to you.

Reply:
No, I think we're good. I was just confused by the ordering. The migration happens before the data is ready, which isn't how I read it the first time through. I think I've got it now and this is ok.